arenarobot/service/processor/apriltag_detector.py

```diff
@@ -0,0 +1,178 @@
+import numpy as np
+import cv2
+from paho.mqtt.client import MQTTMessage
+from transformations import euler_matrix, quaternion_matrix
+from dt_apriltags import Detector
+from arenarobot.service.processor import ArenaRobotServiceProcessor
+from scipy.spatial.transform import Rotation as R
+
+# pylint: disable=too-many-instance-attributes
+class ArenaRobotServiceProcessorApriltagDetector(ArenaRobotServiceProcessor):
+    """Apriltag Detector processor class for the ARENA."""
+
+    DEVICE_INSTANCE_PROCESSOR_TYPE = "apriltag_detector"
+
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-statements
+    def __init__(self, processor_apriltag_detector_topic: str,
+                 processor_apriltag_detector_instance_name: str = None,
+                 scale_factor: int = 1,
+                 camera_resolution = None,
+                 camera_params = None,
+                 dist_params = None,
+                 apriltag_family = None,
+                 apriltag_locations = None,
+                 **kwargs):
+        """Initialize the apriltag detector processor class."""
+        self.processor_apriltag_detector_topic = processor_apriltag_detector_topic
+        self.processor_apriltag_detector_instance_name = processor_apriltag_detector_instance_name
+
+        self.prev_data         = None                  # saving data from previous frame
+        self.cap               = None                  # video capture
+        self.camera_resolution = camera_resolution     # camera resolution [width, height] (pixels)
+        self.mtx               = None                  # camera matrix
+        self.params            = camera_params         # camera matrix parameters fx, fy, cx, cy
+        self.dist_params       = np.array(dist_params) # distortion parameters
+        self.at_detector       = None                  # apriltag detector object
+        self.flip_matrix       = None                  # flip matrix
+        self.apriltag_family   = apriltag_family       # tag family
+        self.apriltags         = apriltag_locations    # known apriltag poses (world coordinates in meters)
+
+        processor_type = (ArenaRobotServiceProcessorApriltagDetector
+                          .DEVICE_INSTANCE_PROCESSOR_TYPE)
+
+        super().__init__(
+            device_instance_processor_type=processor_type,
+            device_instance_prefix="apriltag_detector_",
+            **kwargs
+        )
+    '''
+    Set up Apriltag Detector object and define a bunch of stuff
+    '''
+    def setup(self):
+        # Initialize camera using /dev/video0 (works for USB cameras 
+        # but not Arducam-type cameras because the Raspberry Pi
+        # Foundation is sooooo silly)
+        self.cap = cv2.VideoCapture(0)
+
+        # subsample to increasing processing speed
+        # note: selected resolution affects camera parameters
+        self.cap.set(3, self.camera_resolution[0])
+        self.cap.set(4, self.camera_resolution[1])
```

Comment:
Please change magic numbers to constant in all caps at top of file (outside of class).